### PR TITLE
joi inheritance support

### LINKED ISSUE #2301: another english translation needed
https://usaco.guide/gold/mst?lang=cpp#problem-joi-15-inheritance
bruh

### PR DESCRIPTION
**Pull request by SansPapyrus683**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as
done._

- [x] I have tested my code
- [x] I have followed the code style guidelines mentioned
      [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [x] I have properly formatted the files according to the steps mentioned
      [here](https://usaco.guide/general/adding-solution#steps)

**again, i would like some editorial suggestions, i have no clue how to write proofs**

oh this should close #2301

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
if this doesn't fix the build istg i'm going to barf up literal cacti

File: solutions/joi-15-inheritance.mdx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	vector<Railroad> railroads(rail_num);
+	for (int r = 0; r < rail_num; r++) {
+		Railroad& rr = railroads[r];
+		rr.id = r;
+		std::cin >> rr.city1 >> rr.city2 >> rr.cost;
+		rr.city1--;
+		rr.city2--;
+	}
+	std::sort(railroads.begin(), railroads.end(),
+		[&] (const Railroad& r1, const Railroad& r2) {
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:109]:
```suggestion
		[] (const Railroad& r1, const Railroad& r2) {
```

**Comment by SansPapyrus683** [comment]
ily vercel :heart:

File: solutions/joi-15-inheritance.mdx [hunk 2]
```diff
@@ -0,0 +1,142 @@
+---
+id: joi-15-inheritance
+source: JOI 2015
+title: Inheritance
+author: Kevin Sheng
+---
+
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:7]:
```suggestion

# Explanation

```

File: solutions/joi-15-inheritance.mdx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+However, linearly iterating through the entire list of DSUs would result in a
+runtime of about $\mathcal{O}(KM)$, which is too much given the bounds.
+
+To find the first child that can accept a railroad faster, let's try using binary search!
+But for binary search to actually work, we first need to prove the following:
+
+No matter what railroad we're processing,
+all children after the first child that can accept the current railroad
+can also accept said railroad, while all children *before* that child cannot.
+
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:24]: State this in the form of a lemma (it's unclear what you are trying to prove)

**Review by dutinmeow — CHANGES_REQUESTED** [review]

File: solutions/joi-15-inheritance.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply solutions/joi-15-inheritance.mdx:24]: dustin can you just rewrite this for me

Comment (SansPapyrus683) [reply solutions/joi-15-inheritance.mdx:26]: actually rereading this it's actually ok

File: solutions/joi-15-inheritance.mdx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+			children[valid].link(rr.city1, rr.city2);
+		}
+		ownership[rr.id] = valid + 1;
+	}
+	for (int c : ownership) {
+		cout << c << '\n';
+	}
+}
+```
+
```
Comment (ryanchou-dev) [inline solutions/joi-15-inheritance.mdx:142]:
```suggestion
```

**Review by ryanchou-dev — APPROVED** [review]
love the editorial

**Review by TruaShamu — APPROVED** [review]
Alright, it looks good to me

**Comment by dutinmeow** [comment]
Be consistant with your paragraph spacings

image

**Comment by SansPapyrus683** [comment]
a single newline is for closely related ideas but not closely related enough to be a simple continuation
a new paragraph is for a new section of the editorial

File: solutions/joi-15-inheritance.mdx [hunk 5]
```diff
@@ -0,0 +1,143 @@
+---
+id: joi-15-inheritance
+source: JOI 2015
+title: Inheritance
+author: Kevin Sheng
+---
+
+# Explanation
+
+At first glance, it seems Kruskal's would be more appropriate for this problem, so let's use that.
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:10]:
```suggestion
We will use Kruskal's MST Algorithm to solve this problem.
```

File: solutions/joi-15-inheritance.mdx [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+Let's first give each child their own DSU data structure to store which cities are connected with their railroads,
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:12]: Link DSU Module

File: solutions/joi-15-inheritance.mdx [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+and then iterate through the railroads through decreasing order of price,
+giving each railroad to the first child that can accept it. This is effectively the same as going through
+each child one at a time and constructing a maximum spanning tree for each of them individually.
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:15]: Briefly explain how max spanning tree is equivalent to min span tree

Comment (dutinmeow) [reply solutions/joi-15-inheritance.mdx:15]: *in terms of construction

File: solutions/joi-15-inheritance.mdx [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+However, linearly iterating through the entire list of DSUs would result in a
+runtime of about $\mathcal{O}(KM)$, which is too much given the bounds.
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:18]:
```suggestion
runtime of about $\mathcal{O}(KM)$, which is too slow given the bounds.
```

File: solutions/joi-15-inheritance.mdx [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+To find the first child that can accept a railroad faster, let's try using binary search!
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:20]:
```suggestion
To quickly find the first child that can accept a railroad, we can use binary search. 
```

File: solutions/joi-15-inheritance.mdx [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+But for binary search to actually work, we first need to prove the following:
+
+No matter what railroad we're processing,
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:23]:
```suggestion
**Lemma:** No matter what railroad we're processing,
```

File: solutions/joi-15-inheritance.mdx [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+all children after the first child that can accept the current railroad
+can also accept said railroad, while all children *before* that child cannot.
+
+For this to hold, each DSU's connected components must be obtainable from the one after by adding some edges (0 edges is fine too).
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:27]:
```suggestion
**Proof:** For this to hold, each DSU's connected components must be obtainable from the one after by adding some edges (0 edges is fine too).  
```

File: solutions/joi-15-inheritance.mdx [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+No matter what railroad we're processing,
+all children after the first child that can accept the current railroad
+can also accept said railroad, while all children *before* that child cannot.
+
+For this to hold, each DSU's connected components must be obtainable from the one after by adding some edges (0 edges is fine too).  
+This is obviously true for when we start, so now all we have to prove is that giving a railroad to the first child that can accept it will preserve this relation.
+
+Say we had two children that are adjacent in seniority: $c_i$ and $c_{i+1}$, with $c_i$ being older, and $c_{i+1}$ being the first child that can accept the current railroad.  
+If giving $c_{i+1}$ the railroad would make it so that obtaining $c_i$ from the new $c_{i+1}$ is impossible,
+then that would mean the railroad would connect some two components of $c_i$ and thus also mean that $c_i$ should be the first to accept the railroad, not $c_{i+1}$.
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:32]:
```suggestion
then that would mean the railroad would connect some two components of $c_i$ and thus also mean that $c_i$ should be the first to accept the railroad, not $c_{i+1}$. $\blacksquare$
```
Marks end of proof

File: solutions/joi-15-inheritance.mdx [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+Now that we've finally proved that we can use binary search to find the first child that can accept a railroad,
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:34]:
```suggestion
Using binary search reduces $\mathcal{O}(KM)$ to $\mathcal{O}(K \log M)$, which runs in time. 
```

File: solutions/joi-15-inheritance.mdx [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+we can use it to reduce the factor of $M$ to $\log M$ instead, which makes the algorithm fast enough.
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:35]:
```suggestion
```

File: solutions/joi-15-inheritance.mdx [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+	}
+	std::sort(railroads.begin(), railroads.end(),
+		[] (const Railroad& r1, const Railroad& r2) {
+			return r1.cost > r2.cost;
+	});
+
+	vector<DisjointSets> children(child_num, DisjointSets(city_num));
+	vector<int> ownership(rail_num);
+	for (int r = 0; r < rail_num; r++) {
+		const Railroad& rr = railroads[r];
```
Comment (dutinmeow) [inline solutions/joi-15-inheritance.mdx:118]: No need for a reference here if your not directly modifying. Try to keep the code as simple as possible

**Comment by dutinmeow** [comment]
> a single newline is for closely related ideas but not closely related enough to be a simple continuation a new paragraph is for a new section of the editorial

From a reader's POV it just looks like its misindented.

**Comment by SansPapyrus683** [comment]
well you tell me how i do it then

File: solutions/joi-15-inheritance.mdx (hunk 15 above)
Comment (SansPapyrus683) [reply solutions/joi-15-inheritance.mdx:118]: i mean does it really matter

File: solutions/joi-15-inheritance.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/joi-15-inheritance.mdx:10]: this implies prim works too

File: solutions/joi-15-inheritance.mdx (hunk 7 above)
Comment (SansPapyrus683) [reply solutions/joi-15-inheritance.mdx:15]: that's... kinda obvious ngl

File: solutions/joi-15-inheritance.mdx (hunk 12 above)
Comment (SansPapyrus683) [reply solutions/joi-15-inheritance.mdx:32]: haven't seen that anywhere else

File: solutions/joi-15-inheritance.mdx (hunk 10 above)
Comment (SansPapyrus683) [reply solutions/joi-15-inheritance.mdx:23]: hold on
aren't lemma's supposed to be part of a bigger proof?
what's the "bigger proof" then

**Review by dutinmeow — APPROVED** [review]

**Pull request merged**